### PR TITLE
Remove null-return on preg_replace() with valid patterns

### PR DESCRIPTION
**Pull request by staabm**
~Remove null-return on preg_replace() with valid patterns~
~Return benevolent union from preg_replace()~

closes https://github.com/phpstan/phpstan/issues/11547

requires https://github.com/phpstan/phpstan-src/pull/3339

### CODE REVIEW

**Comment by staabm** [comment]
//cc @thg2k

**Comment by thg2k** [comment]
See the original issue. I was proven wrong: a perfectly valid regexp can still fail based on the input. So maybe if the regexp looks correct we return a benevolent union with null?

**Comment by staabm** [comment]
failing build in phpdoc-parser repo is related to a baselined error which no longer happens after the PR (because benevolent)

File: tests/PHPStan/Analyser/nsrt/non-empty-string-replace-functions.php [hunk 1]
```diff
@@ -18,7 +18,7 @@ public function replace(string $search, string $replacement, string $subject){
         assertType('non-empty-string', str_replace($search, $replacement, $subject));
         assertType('non-empty-string', str_ireplace($search, $replacement, $subject));
 
-        assertType('non-empty-string|null', preg_replace($search, $replacement, $subject));
+        assertType('(non-empty-string|null)', preg_replace($search, $replacement, $subject));
```
Comment (thg2k) [inline tests/PHPStan/Analyser/nsrt/non-empty-string-replace-functions.php]: Seems a preexisting problem, but it's wrong: `preg_replace('/^a(b?)/', '\\1', 'a');`

Comment (staabm) [reply tests/PHPStan/Analyser/nsrt/non-empty-string-replace-functions.php]: Please open a new issue

**Comment by ondrejmirtes** [comment]
Do we want this? Why it would be benevolent? Do we not want the user to always check for errors?

**Comment by staabm** [comment]
I think it was one of the requests of the initial issue. dropped benevolence now and reduced it to the bugfix only

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**